drivers/wifi/esp8266/esp8266.c

```diff
@@ -0,0 +1,784 @@
+	struct esp8266_data *data = dev->driver_data;
+	int ret, len;
+
+	if (!net_if_is_up(data->net_iface)) {
+		return -EIO;
+	}
+
+	if (esp8266_flag_is_set(data, EDF_STA_CONNECTED) ||
+	    esp8266_flag_is_set(data, EDF_STA_CONNECTING)) {
+		return -EALREADY;
+	}
+
+	esp8266_flag_set(data, EDF_STA_CONNECTING);
+
+	len = snprintk(cmd, sizeof(cmd), "AT+CWJAP_CUR=\"");
+	memcpy(&cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+		memcpy(&cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+	}
+
+	len += snprintk(&cmd[len], sizeof(cmd) - len, "\"");
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, cmd, &data->sem_response,
+			     K_NO_WAIT);
+
+	esp8266_flag_clear(data, EDF_STA_CONNECTING);
+
+	return ret;
+}
+
+static int esp8266_mgmt_disconnect(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWQAP", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_enable(struct device *dev,
+				  struct wifi_connect_req_params *params)
+{
+	char cmd[sizeof("AT+CWSAP_CUR=\"\",\"\",xx,x") + WIFI_SSID_MAX_LEN +
+		 WIFI_PSK_MAX_LEN];
+	struct esp8266_data *data = dev->driver_data;
+	int ecn = 0, len, ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWMODE_CUR=3", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable AP mode, ret %d", ret);
+		return ret;
+	}
+
+	len = snprintk(cmd, sizeof(cmd), "AT+CWSAP_CUR=\"");
+	memcpy(&cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+		memcpy(&cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+		ecn = 3;
+	} else {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+	}
+
+	snprintk(&cmd[len], sizeof(cmd) - len, "\",%d,%d", params->channel,
+		 ecn);
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, cmd, &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_disable(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWMODE_CUR=1", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static void esp8266_init_work(struct k_work *work)
+{
+	struct esp8266_data *dev;
+	int ret;
+	static struct setup_cmd setup_cmds[] = {
+		/* turn off echo */
+		SETUP_CMD_NOHANDLE("ATE0"),
+		/* enable multiple socket support */
+		SETUP_CMD_NOHANDLE("AT+CIPMUX=1"),
+		SETUP_CMD_NOHANDLE("AT+CWMODE_CUR=1"),
```

Comment:
With this AT command could you try "AT+CWMODE=1" I believe the firmware of the 8266 supports this. The manuals say CWMODE_CUR but I believe it will work, but I don't have a esp8266 to try.

Reply:
I will change it. The reason for the '_CUR' commands is that I've actually been using the NONOS_SDK (https://github.com/espressif/ESP8266_NONOS_SDK) which has some minor differences in the AT command set but I've seen that it is actually end-of-life now and the RTOS_SDK (https://github.com/espressif/ESP8266_RTOS_SDK) is the recommended one.

I will change to follow these AT commands: https://github.com/espressif/esp-at/blob/master/docs/ESP_AT_Commands_Set.md

If I understand correctly, it should just work with ESP32 as well without any changes. I'll post an update tomorrow.

Reply:
I found that earlier today whilst searching the espressif forums.

Maybe you should rename to a ESP_WIFI_MODEM ?

My only concern is will there be issues with older firmware?

Reply:
> Maybe you should rename to a ESP_WIFI_MODEM ?

Yea that sounds like a good idea.

> My only concern is will there be issues with older firmware?

Yup. I'm not really sure how to tackle it but I think that it might actually be okay to not work with older firmwares as long as it is clearly stated which ones are supported. One could even add the `AT+GMR` command to get the AT and SDK versions.

Reply:
Using GMR to get the firmware version had crossed my mind but it may not be neccassary; 
I just had a look at the 'latest' eps8266 command examples manual (dated 2017) and its very similar to esp32. I'm going to be as bold as to say we should be fine with a singular esp AT WiFi driver.

**Espressif Doc Library:**
https://www.espressif.com/en/support/download/documents?keys=&field_type_tid%5B%5D=14

Reply:
I added a choice in the config for selecting AT 1.7 or AT 2.0. For ESP32 I believe you should select AT 2.0.